File: ttnn/ttnn/operations/downsample.py
# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

from typing import Tuple


import ttnn
import torch


def _golden_function(
    input_tensor: ttnn.Tensor,
    downsample_params: Tuple[float, float, float, float, float],
    output_dtype: ttnn.DataType,
    **_,
):
    input_tensor = input_tensor.permute(0, 3, 1, 2)
    ret = torch.nn.functional.downsample(input_tensor, downsample_params=downsample_params, dtype=dtype)
    ret = ret.permute(0, 2, 3, 1)
    return ret


downsample = ttnn.attach_golden_function(ttnn.downsample, golden_function=_golden_function)

__all__ = []

File: ttnn/cpp/ttnn/operations/pool/downsample/downsample.cpp
// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
//
// SPDX-License-Identifier: Apache-2.0

#include "downsample.hpp"

#include <boost/container/vector.hpp>
#include <algorithm>
#include <vector>

#include "device/downsample_op.hpp"
#include <tt-metalium/buffer_constants.hpp>
#include "ttnn/run_operation.hpp"
#include "ttnn/tensor/types.hpp"

using namespace tt::tt_metal;

namespace ttnn::operations::downsample {

Tensor ExecuteDownsample::invoke(
    const Tensor& input_tensor_a, std::array<uint32_t, 5> downsample_params, std::optional<DataType> dtype) {
    auto dtype_ = dtype.has_value() ? dtype.value() : input_tensor_a.get_dtype();
    auto output_tensor = operation::run(Downsample{downsample_params, dtype_}, {input_tensor_a}).front();
    return output_tensor;
};

}  // namespace ttnn::operations::downsample
